src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -271,30 +284,112 @@ type Levels m h = V.Vector (Level m h)
+      -- completed. If 'MergeKnownCompleted', then we are sure the merge has been
+      -- completed, otherwise if 'MergeMaybeCompleted' we have to check the
       -- 'MergingRunState'.
-      !(StrictMVar m (MergingRunState m h))
-    -- | The result of merging a single run, is a single run.
-  | SingleRun !(Run m h)
+    , mergeKnownCompleted :: !(MutVar (PrimState m) MergeKnownCompleted)
+    , mergeState          :: !(StrictMVar m (MergingRunState m h))
+    , mergeRefCounter     :: !(RC.RefCounter m)
+    }
+
+{-# SPECIALISE newMergingRun ::
+     MergePolicyForLevel
+  -> NumRuns
+  -> NumEntries
+  -> MergeKnownCompleted
+  -> MergingRunState IO h
+  -> IO (MergingRun IO h)
+  #-}
+-- This allows constructing ill-formed MergingRuns, but the flexibility is
+-- needed for creating a merging run that is already Completed, as well as
+-- opening a merging run from a snapshot.
+-- TODO: instead create a Single run when OneShot merging?
+-- TODO: do not store MergeKnownCompleted in snapshot? It's redundant.
+-- TODO: slightly different API for opening from snapshot?
+newMergingRun ::
+     (MonadMVar m, MonadMask m, MonadSTM m, MonadST m)
+  => MergePolicyForLevel
+  -> NumRuns
+  -> NumEntries
+  -> MergeKnownCompleted
+  -> MergingRunState m h
+  -> m (MergingRun m h)
+newMergingRun mergePolicy mergeNumRuns mergeNumEntries knownCompleted state = do
+    mergeUnspentCredits <- UnspentCreditsVar <$> newPrimVar 0
+    mergeStepsPerformed <- TotalStepsVar <$> newPrimVar 0
+    case state of
+      OngoingMerge{}   -> assert (knownCompleted == MergeMaybeCompleted) (pure ())
+      CompletedMerge{} -> pure ()
+    mergeKnownCompleted <- newMutVar knownCompleted
+    mergeState <- newMVar $! state
+    mergeRefCounter <- RC.mkRefCounter1 (Just (finalise mergeState))
+    pure $! MergingRun {
+        mergePolicy
+      , mergeNumRuns
+      , mergeNumEntries
+      , mergeUnspentCredits
+      , mergeStepsPerformed
+      , mergeKnownCompleted
+      , mergeState
+      , mergeRefCounter
+      }
+  where
+    finalise var = withMVar var $ \case
+        CompletedMerge r ->
+          Run.removeReference r
+        OngoingMerge rs _ m -> do
+          V.forM_ rs Run.removeReference
+          Merge.abort m
+
+-- | Create references to the runs that should be queried for lookups.
+-- In particular, if the merge is not complete, these are the input runs.
+mergingRunGetRuns ::
+     (PrimMonad m, MonadMVar m, MonadMask m)
+  => TempRegistry m
+  -> MergingRun m h
+  -> m (V.Vector (Run m h))
+mergingRunGetRuns reg mr =
+    -- We take the references while holding the MVar to make sure the MergingRun
+    -- does not get completed concurrently before we are done.
+    withMVar (mergeState mr) $ \case
+      CompletedMerge r -> V.singleton <$> allocRun r
+      OngoingMerge rs _ _ -> rs <$ V.forM_ rs allocRun
+  where
+    allocRun r = do
+        allocateTemp reg (Run.addReference r) (\_ -> Run.removeReference r)
+        pure r
```

Comment:
I think that the name is not clear about the fact that it adds references. Maybe we should call it `tempAddReferenceInputRuns` or something along those lines?

Reply:
Naming is difficult here. It also doesn't just add references, and it can also return the resulting run (instead of input runs). 

Once it's in a `MergingRun` module and returns `Vector (Ref (Run m h))` I would call it something like `{get,make,create,clone}RunRefs`, depending on terminology used in the `Ref` API. Maybe `mergingRunCloneRunRefs` is already a decent choice now?